### DISCUSSION TITLE
Make FUSE mount in docker container available to other containers/host

### DISCUSSION THREAD

**Discussion by r-ash** [seaweedfs/seaweedfs#2114] [Q&A]
Is it possible to expose FUSE mount running inside a docker container to outside file system / other containers?

Reproducible example below

```
docker network create seaweed-network
docker run -d --rm -p 9333:9333 --network seaweed-network --name master chrislusf/seaweedfs master -ip=master
docker run -d --rm -p 8080:8080 -p 18080:18080 --network seaweed-network --name volume chrislusf/seaweedfs volume -max=5 -mserver="master:9333" -port=8080
docker run -d --rm --name filer -p 8888:8888 -p 18888:18888 --network seaweed-network chrislusf/seaweedfs filer -port=8888 -master="master:9333"
mkdir test
docker run -d --rm --name mount --network seaweed-network --mount type=bind,src="$(pwd)"/test,dst=/vol,bind-propagation=rshared --device /dev/fuse --privileged chrislusf/seaweedfs mount -filer=filer:8888 -dir=/vol -filer.path=/
echo "test file" > test/test.txt
curl -H "Accept: application/json" http://localhost:8888  ## Expecting this to list the file but the volume is empty
docker exec mount ls /vol
```
What I am expecting here is that the test file `test.txt` will be stored in SeaweedFS volume and available in the `mount` container. Basing this off suggestion on https://stackoverflow.com/questions/53631567/share-a-fuse-fs-mounted-inside-a-docker-container-through-volumes that bind mounting the container should make this available. But perhaps something about how FUSE mount is working makes this not possible?

For context what I want to achieve is running two containers, A which writes data into a docker volume and B which reads data from volume and does some further processing. Trying to work out how I can do processing on B with SeaweedFS as the storage so this can be deployed on a different machine.

Thanks!

**Comment by chrislusf** [comment]
> For context what I want to achieve is running two containers, A which writes data into a docker volume and B which reads data from volume and does some further processing.

This is achieved, correct?

> Trying to work out how I can do processing on B with SeaweedFS as the storage so this can be deployed on a different machine.

Mount SeaweedFS on the different machine?

**Reply by r-ash** [reply]
> > For context what I want to achieve is running two containers, A which writes data into a docker volume and B which reads data from volume and does some further processing.
> 
> This is achieved, correct?

I think this isn't at the moment (and perhaps is not possible at all?) Example below just has the case of writing data from container A 
```
docker run -d --rm -p 9333:9333 --network seaweed-network --name master chrislusf/seaweedfs master -ip=master
docker run -d --rm -p 8080:8080 -p 18080:18080 --network seaweed-network --name volume chrislusf/seaweedfs volume -max=5 -mserver="master:9333" -port=8080
docker run -d --rm --name filer -p 8888:8888 -p 18888:18888 --network seaweed-network chrislusf/seaweedfs filer -port=8888 -master="master:9333"
mkdir test
docker run -d --rm --name mount --network seaweed-network --mount type=bind,src="$(pwd)"/test,dst=/vol,bind-propagation=rshared --device /dev/fuse --privileged chrislusf/seaweedfs mount -filer=filer:8888 -dir=/vol -filer.path=/
docker run -d --rm --name containerA --mount type=bind,src="$(pwd)"/test,dst=/vol,bind-propagation=rshared alpine tail -f /dev/null
docker exec containerA ls /vol
docker exec containerA sh -c "echo 'Sample file' > /vol/sample_file.txt"
curl -H "Accept: application/json" http://localhost:8888/sample_file.txt     ## Returns no data
```
Containers running a Seaweed master, volume, filer and mount. And a container `containerA` which has some process which writes data to disk. Where the mount & container A are bind mounted to host file system at same location. 

What I am hoping here is that bind mounting the docker container running the FUSE mount will mean that writes from other containers also bind mounted to the same dir will be able to use the FUSE mount. It looks like that doesn't work, any file written by container A reflects on host by isn't stored in SeaweedFS (or available in the FUSE mount container, `docker exec mount ls /vol` is empty)

**Reply by chrislusf** [reply]
If you want to read/write data from containerA, you can `weed mount` from containerA.

**Reply by r-ash** [reply]
Thanks for your help, that sounds like a good approach.

**Reply by chrislusf** [reply]
Sure. BTW: You are working on "infectious disease epidemiology". Is it COVID19? :)

**Reply by r-ash** [reply]
I'm not personally doing any covid work, mostly HIV. But we're hoping to use this as a storage backend for a general tool https://github.com/mrc-ide/rrq which we use in the department over many projects (including the covid response).

**Reply by chrislusf** [reply]
seems not huge amount of data for rrq. What is the reason to use SeaweedFS?

**Reply by r-ash** [reply]
We've developed a web app which essentially makes a model of HIV available to non-technical users. So users upload data then run the model which saves outputs to disk which they can download later. The input and output files can be several MB. Model runs can take ~30 mins or longer depending on the country and so we use rrq to queue users and make their results available at later point. This works fine most times just saving model inputs and outputs to disk but the usage of the app very spiky. We have very little load throughout the year but there are yearly workshops run where all countries try to produce their estimates at the same time. So we'd like to be able to spin up additional rrq workers on other machines but can't atm due to them all needing access to the same file system.
